implements FunctionArgumentsSpacingRule

u-abyss opened the pull request:
closed #5224

implements `functions_arguments_spacing` rule to detects unnecessary spaces before the first argument of a function and unnecessary spaces in the last argument.

nonTriggeringExamples:
```
makeGenerator()
makeGenerator(style)
makeGenerator(true, false)
```

triggeringExamples
```
makeGenerator(↓ style)
makeGenerator(style ↓)
makeGenerator(↓ style ↓)
makeGenerator(↓ offset: 0, limit: 0)
makeGenerator(offset: 0, limit: 0 ↓)
makeGenerator(↓ 1, 2, 3 ↓)
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,38 @@
+        identifier: "functions_arguments_spacing",
+        name: "Function Arguments Spacing",
+        description: "Remove the space before the first function argument and after the last argument",
+        kind: .lint
```
Please move the test examples into the rule description. They are automatically executed as test and serve documentation purposes.

SimplyDanny commented on Tests/SwiftLintFrameworkTests/CyclomaticComplexityRuleTests.swift:
```diff
@@ -63,7 +63,6 @@ class CyclomaticComplexityRuleTests: SwiftLintTestCase {
 
         let description = baseDescription.with(nonTriggeringExamples: nonTriggeringExamples)
                                          .with(triggeringExamples: triggeringExamples)
-
```
Unrelated.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,38 @@
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override func visitPost(_ node: FunctionCallExprSyntax) {
+            let argsCount = node.arguments.count
+            guard argsCount != 0 else {
```
The case of no parameters could also be interesting if you consider for instance `f( )`.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,38 @@
+
+            let leftParanTrailingTrivia = node.leftParen?.trailingTrivia
+            if leftParanTrailingTrivia == Trivia.space {
+                violations.append(node.leftParen!.endPositionBeforeTrailingTrivia)
```
We avoid force-unwrapping. Please use optional bindings instead.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,38 @@
+                return
+            }
+
+            let leftParanTrailingTrivia = node.leftParen?.trailingTrivia
```
It's sometimes a bit surprising to which node SwiftSyntax attaches the whitespace. Therefore, you should also check the leading trivia of the first argument and the trailing trivia of the last argument.

SimplyDanny reviewed (changes requested):
Thanks for taking this up!

Please add a changelog entry to let folks now about the new rule and to credit yourself.

We should also carefully review the reported OSS findings. Often, they disclose case that we haven't thought of just yet.

u-abyss commented:
@SimplyDanny 
Thank you for your kind reviews.
The CI is currently failing and I would like to know why it is failing as I cannot find the cause on my end.
I have confirmed success by running the swift test locally.

SimplyDanny commented:
> @SimplyDanny Thank you for your kind reviews. The CI is currently failing and I would like to know why it is failing as I cannot find the cause on my end. I have confirmed success by running the swift test locally.

Please rebase your branch onto the current `main`. That should resolve the test failures.

u-abyss commented:
@SimplyDanny 
I rebased my branch onto current main branch, but CI still fail...

SimplyDanny commented:
> @SimplyDanny 
> 
> I rebased my branch onto current main branch, but CI still fail...

Merge is not the same as rebase. 😉

Danger fails because of merge commits which are undesirable.

u-abyss commented:
@SimplyDanny 
It was not rebased correctly; rebasing my branch onto the current main branch, and the CI passed.
thanks.
I fixed the parts that were reviewed.
I've added some test cases, but I'm not sure I've covered all of them, so if there are any missing, please point them out!

SimplyDanny commented on Tests/SwiftLintFrameworkTests/FunctionArgumentsSpacingRuleTests.swift:
This test can be removed entirely. Examples in rule descriptions will be verified in any case.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,93 @@
+            // as it may contain trivia that is not space like blockComment(_:)
+            if let firstArgumentLeadingTrivia = firstArgument?.leadingTrivia,
+               !firstArgumentLeadingTrivia.containsNewlines() {
+                    if let firstElementTrivia = firstArgumentLeadingTrivia.reversed().first {
```
```suggestion
                    if let firstElementTrivia = firstArgumentLeadingTrivia.pieces.last {
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,93 @@
+}
+
+private extension TriviaPiece {
+    var isSpaces: Bool {
```
Should this also check for tabs `\t`?

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,93 @@
+            """)
+        ],
+        triggeringExamples: [
+            Example("makeGenerator(↓ )"),
```
You may choose a shorter name for the function like just `f`. `makeGenerator` seems to imply a special meaning which it hasn't.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/FunctionArgumentsSpacingRule.swift:
```diff
@@ -0,0 +1,93 @@
+    static let description = RuleDescription(
+        identifier: "functions_arguments_spacing",
+        name: "Function Arguments Spacing",
+        description: "Remove the space before the first function argument and after the last argument",
```
I wonder if this rule should do more than checking the first and the last argument. Can it check all arguments in a function call, so that there is no whitespace before and first and after the last argument and exactly one space after every comma?

u-abyss commented:
@SimplyDanny 
The implementation has been changed to ensure that there is not only no space before the first argument and after the last argument, but that there is only one space after the comma in all arguments.
For comments, the implementation has been changed to allow one space before the comment when a comment is written after  an argument or a comma.
